<!DOCTYPE html>
<html lang="en-us">

<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="generator" content="Source Themes Academic 4.3.1">

  

  
  
  
  
  
    
    
    
  
  

  <meta name="author" content="Nicholas Ollberding">

  
  
  
    
  
  <meta name="description" content="This post is from a tutorial demonstrating the processing of amplicon short read data in R taught as part of the Introduction to Metagenomics Summer Workshop. It provides a quick introduction some of the functionality provided by phyloseq and follows some of Paul McMurdie’s excellent tutorials. This tutorial picks up where Ben Callahan’s DADA2 tutorial leaves off and highlights some of the main accessor and processor functions of the package.">

  
  <link rel="alternate" hreflang="en-us" href="/post/introduction-to-phyloseq/">

  


  

  
  
  
  <meta name="theme-color" content="#2962ff">
  

  
  
  
  
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.6/css/academicons.min.css" integrity="sha256-uFVgMKfistnJAfoCUQigIl+JfUaP47GrRKjf6CTPVmw=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.6.0/css/all.css" integrity="sha384-aOkxzJ5uQz7WBObEZcHvV5JvRW3TUc2rNPA7pe3AwnsUohiw1Vj2Rgx2KSOkF5+h" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">

    
    
    
      
    
    
      
      
        
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/styles/github.min.css" crossorigin="anonymous" title="hl-light">
          <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/styles/dracula.min.css" crossorigin="anonymous" title="hl-dark" disabled>
        
      
    

    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.css" integrity="sha512-M2wvCLH6DSRazYeZRIm1JnYyh22purTM+FDB5CsyxtQJYeKq83arPe5wgbNmcFXGqiSH2XR8dT/fJISVA1r/zQ==" crossorigin="anonymous">
    

    

  

  
  
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat:400,700|Roboto:400,400italic,700|Roboto+Mono&display=swap">
  

  
  
  
  <link rel="stylesheet" href="/css/academic.min.3bc694af15fd1e4ff7262e7dd46f11a8.css">

  

  
  
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-144821860-1', 'auto');
      
      ga('require', 'eventTracker');
      ga('require', 'outboundLinkTracker');
      ga('require', 'urlChangeTracker');
      ga('send', 'pageview');
    </script>
    <script async src="https://www.google-analytics.com/analytics.js"></script>
    
    <script async src="https://cdnjs.cloudflare.com/ajax/libs/autotrack/2.4.1/autotrack.js" integrity="sha512-HUmooslVKj4m6OBu0OgzjXXr+QuFYy/k7eLI5jdeEy/F4RSgMn6XRWRGkFi5IFaFgy7uFTkegp3Z0XnJf3Jq+g==" crossorigin="anonymous"></script>
    
  
  

  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/introduction-to-phyloseq/">

  
  
  
  
    
    
  
  <meta property="twitter:card" content="summary">
  
  <meta property="og:site_name" content="">
  <meta property="og:url" content="/post/introduction-to-phyloseq/">
  <meta property="og:title" content="Introduction to Phyloseq | ">
  <meta property="og:description" content="This post is from a tutorial demonstrating the processing of amplicon short read data in R taught as part of the Introduction to Metagenomics Summer Workshop. It provides a quick introduction some of the functionality provided by phyloseq and follows some of Paul McMurdie’s excellent tutorials. This tutorial picks up where Ben Callahan’s DADA2 tutorial leaves off and highlights some of the main accessor and processor functions of the package."><meta property="og:image" content="/img/icon-192.png">
  <meta property="twitter:image" content="/img/icon-192.png"><meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2019-07-28T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2019-07-28T22:23:51-04:00">
  

  


  





  <title>Introduction to Phyloseq | </title>

</head>

<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" >

  <aside class="search-results" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search">
        
      </div>

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>


  
<nav class="navbar navbar-light fixed-top navbar-expand-lg py-0 compensate-for-scrollbar" id="navbar-main">
  <div class="container">

    
      <a class="navbar-brand" href="/"></a>
      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
        <span><i class="fas fa-bars"></i></span>
      </button>
      

    
    <div class="collapse navbar-collapse" id="navbar">

      
      
      <ul class="navbar-nav mr-auto">
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#about"><span>Home</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#posts"><span>Posts</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#projects"><span>Projects</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#featured"><span>Publications</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        

        <li class="nav-item">
          <a class="nav-link " href="/courses/"><span>Courses</span></a>
        </li>

        
        

        

        
        
        
          
        

        
        
        
        
        
        
          
          
          
            
          
          
        

        <li class="nav-item">
          <a class="nav-link " href="/#contact"><span>Contact</span></a>
        </li>

        
        

      
      </ul>
      <ul class="navbar-nav ml-auto">
      

        

        
        <li class="nav-item">
          <a class="nav-link js-search" href="#"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        

        
        <li class="nav-item">
          <a class="nav-link js-dark-toggle" href="#"><i class="fas fa-moon" aria-hidden="true"></i></a>
        </li>
        

      </ul>

    </div>
  </div>
</nav>


  <article class="article" itemscope itemtype="http://schema.org/Article">

  












  

  
  
  
<div class="article-container pt-3">
  <h1 itemprop="name">Introduction to Phyloseq</h1>

  

  
    



<meta content="2019-07-28 00:00:00 &#43;0000 UTC" itemprop="datePublished">
<meta content="2019-07-28 22:23:51 -0400 EDT" itemprop="dateModified">

<div class="article-metadata">

  
  

  
  <span class="article-date">
    
    
      
          Last updated on
      
    
    <time>Jul 28, 2019</time>
  </span>
  

  

  
  <span class="middot-divider"></span>
  <span class="article-reading-time">
    18 min read
  </span>
  

  
  

  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fas fa-folder"></i>
    <a href="/categories/microbiome/">Microbiome</a></span>
  

  
    
<div class="share-box" aria-hidden="true">
  <ul class="share">
    
      
      
      
        
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=/post/introduction-to-phyloseq/&amp;text=Introduction%20to%20Phyloseq" target="_blank" rel="noopener" class="share-btn-twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=/post/introduction-to-phyloseq/&amp;t=Introduction%20to%20Phyloseq" target="_blank" rel="noopener" class="share-btn-facebook">
          <i class="fab fa-facebook-f"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="mailto:?subject=Introduction%20to%20Phyloseq&amp;body=/post/introduction-to-phyloseq/" target="_blank" rel="noopener" class="share-btn-email">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=/post/introduction-to-phyloseq/&amp;title=Introduction%20to%20Phyloseq" target="_blank" rel="noopener" class="share-btn-linkedin">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://web.whatsapp.com/send?text=Introduction%20to%20Phyloseq%20/post/introduction-to-phyloseq/" target="_blank" rel="noopener" class="share-btn-whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=/post/introduction-to-phyloseq/&amp;title=Introduction%20to%20Phyloseq" target="_blank" rel="noopener" class="share-btn-weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>


  

</div>

    














  
</div>



  <div class="article-container">

    <div class="article-style" itemprop="articleBody">
      


<div id="section" class="section level1">
<h1></h1>
<p><br></p>
<p>This post is from a tutorial demonstrating the processing of amplicon short read data in R taught as part of the <a href="https://github.com/Nick243/Introduction-to-Metagenomics-Summer-Workshop-2019">Introduction to Metagenomics Summer Workshop</a>. It provides a quick introduction some of the functionality provided by phyloseq and follows some of Paul McMurdie’s excellent tutorials. This tutorial picks up where Ben Callahan’s DADA2 tutorial leaves off and highlights some of the main accessor and processor functions of the package. I thought it might be useful to a broader audience so decided to post it.</p>
<p><br></p>
<p>The goal of this interactive session is to <strong>introduce you to some of the basic functionality of the</strong> <a href="https://www.bioconductor.org/packages/release/bioc/html/phyloseq.html">phyloseq package</a> that can help you to explore and better understand your metagenomic data. We will be working with the phyloseq object that was created during the <a href="http://benjjneb.github.io/dada2/tutorial.html">DADA2 tutorial</a>. If you recall, these were murine stool samples collected from a single mouse over time. The phyloseq object contains: an ASV table, sample metadata, taxonomic classifications, and the reference sequences. We did not generate a phylogenetic tree from these sequences, but if we had, it could be included as well.</p>
<p>The session will quickly cover some of the basic accessor, analysis and graphical functions available to you when using the phyloseq package in R.</p>
<p>To learn more, Paul McMurdie has an excellent set of tutorials that I encourage you to explore.</p>
<ul>
<li><a href="https://joey711.github.io/phyloseq/preprocess.html" class="uri">https://joey711.github.io/phyloseq/preprocess.html</a></li>
<li><a href="https://joey711.github.io/phyloseq/index.html" class="uri">https://joey711.github.io/phyloseq/index.html</a></li>
</ul>
<p><br></p>
</div>
<div id="loading-required-packages-and-phyloseq-object" class="section level1">
<h1>Loading required packages and phyloseq object</h1>
<pre class="r"><code>library(dada2); packageVersion(&quot;dada2&quot;)           </code></pre>
<pre><code>## Loading required package: Rcpp</code></pre>
<pre><code>## [1] &#39;1.12.1&#39;</code></pre>
<pre class="r"><code>library(phyloseq); packageVersion(&quot;phyloseq&quot;)      </code></pre>
<pre><code>## [1] &#39;1.28.0&#39;</code></pre>
<pre class="r"><code>library(ggplot2); packageVersion(&quot;ggplot2&quot;)        </code></pre>
<pre><code>## [1] &#39;3.2.0&#39;</code></pre>
<p><br></p>
<p>If the phyloseq (ps) object is not already loaded into your environment…let’s go ahead and do that now. You will need to change the path so that it maps to the ps object on your computer. No path is needed if you are working in an RStudio project folder (or if you cloned the folder from GitHub).</p>
<pre class="r"><code>ps &lt;- readRDS(&quot;C:/Users/olljt2/Desktop/academic_web_page/static/data/ps.rds&quot;)</code></pre>
<p><br></p>
</div>
<div id="accessing-the-sample-information-and-sample-metadata" class="section level1">
<h1>Accessing the sample information and sample metadata</h1>
<pre class="r"><code>ps</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 232 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 232 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 232 reference sequences ]</code></pre>
<ul>
<li>Here we can see that we have a phyloseq object that consists of:
<ul>
<li>An OTU table with 232 taxa and 19 samples</li>
<li>A sample metadata file consisting of 4 variables</li>
<li>A taxonomy table with 7 ranks</li>
<li>Reference sequences on all 232 taxa</li>
</ul></li>
</ul>
<p><br></p>
<p>This highlights one of the key advantages of working with phyloseq objects in R. Each of these data structures is contained in a single object. This makes it easy to keep all of your data together and to share it with colleagues or include it as a supplemental file to a publication.</p>
<p>Next we will see how each of the components can be accessed. We will run through several commands below and then discuss the output.</p>
<pre class="r"><code>nsamples(ps)</code></pre>
<pre><code>## [1] 19</code></pre>
<pre class="r"><code>sample_names(ps)</code></pre>
<pre><code>##  [1] &quot;F3D0&quot;   &quot;F3D1&quot;   &quot;F3D141&quot; &quot;F3D142&quot; &quot;F3D143&quot; &quot;F3D144&quot; &quot;F3D145&quot;
##  [8] &quot;F3D146&quot; &quot;F3D147&quot; &quot;F3D148&quot; &quot;F3D149&quot; &quot;F3D150&quot; &quot;F3D2&quot;   &quot;F3D3&quot;  
## [15] &quot;F3D5&quot;   &quot;F3D6&quot;   &quot;F3D7&quot;   &quot;F3D8&quot;   &quot;F3D9&quot;</code></pre>
<pre class="r"><code>sample_variables(ps)</code></pre>
<pre><code>## [1] &quot;Subject&quot; &quot;Gender&quot;  &quot;Day&quot;     &quot;When&quot;</code></pre>
<pre class="r"><code>head(sample_data(ps))</code></pre>
<pre><code>##        Subject Gender Day  When
## F3D0         3      F   0 Early
## F3D1         3      F   1 Early
## F3D141       3      F 141  Late
## F3D142       3      F 142  Late
## F3D143       3      F 143  Late
## F3D144       3      F 144  Late</code></pre>
<pre class="r"><code>sample_data(ps)$When</code></pre>
<pre><code>##  [1] &quot;Early&quot; &quot;Early&quot; &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot; 
##  [9] &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot;  &quot;Late&quot;  &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot;
## [17] &quot;Early&quot; &quot;Early&quot; &quot;Early&quot;</code></pre>
<pre class="r"><code>table(sample_data(ps)$When)</code></pre>
<pre><code>## 
## Early  Late 
##     9    10</code></pre>
<pre class="r"><code>median(sample_data(ps)$Day)</code></pre>
<pre><code>## [1] 141</code></pre>
<pre class="r"><code>metadata &lt;- data.frame(sample_data(ps))
head(metadata)</code></pre>
<pre><code>##        Subject Gender Day  When
## F3D0         3      F   0 Early
## F3D1         3      F   1 Early
## F3D141       3      F 141  Late
## F3D142       3      F 142  Late
## F3D143       3      F 143  Late
## F3D144       3      F 144  Late</code></pre>
<p>Here we can see that we have 19 samples and they are assigned the sample names we gave them during the DADA2 tutorial.
We also have 4 variables (Subject, Gender, Day, and When) and that information can be easily accessed and computations or descriptive statistics performed.
Specific components of the ps object can be extracted and converted to a data.frame for additional analyses.</p>
<p><br></p>
</div>
<div id="examining-the-number-of-reads-for-each-sample" class="section level1">
<h1>Examining the number of reads for each sample</h1>
<p>Phyloseq makes it easy to calculate the total number of reads for each sample, sort them to identify potentially problematic samples, and plot their distribution.</p>
<pre class="r"><code>sample_sums(ps)</code></pre>
<pre><code>##   F3D0   F3D1 F3D141 F3D142 F3D143 F3D144 F3D145 F3D146 F3D147 F3D148 
##   6528   5017   4863   2521   2518   3488   5820   3879  13006   9935 
## F3D149 F3D150   F3D2   F3D3   F3D5   F3D6   F3D7   F3D8   F3D9 
##  10653   4240  16835   5491   3716   6679   4217   4547   6015</code></pre>
<pre class="r"><code>sort(sample_sums(ps))</code></pre>
<pre><code>## F3D143 F3D142 F3D144   F3D5 F3D146   F3D7 F3D150   F3D8 F3D141   F3D1 
##   2518   2521   3488   3716   3879   4217   4240   4547   4863   5017 
##   F3D3 F3D145   F3D9   F3D0   F3D6 F3D148 F3D149 F3D147   F3D2 
##   5491   5820   6015   6528   6679   9935  10653  13006  16835</code></pre>
<pre class="r"><code>hist(sample_sums(ps), main=&quot;Histogram: Read Counts&quot;, xlab=&quot;Total Reads&quot;, 
     border=&quot;blue&quot;, col=&quot;green&quot;, las=1, breaks=12)</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/number%20of%20reads-1.png" width="672" /></p>
<pre class="r"><code>metadata$total_reads &lt;- sample_sums(ps)</code></pre>
<p>Here we see that the number of reads per sample ranges from 2,518 to 16,835 and most samples have less than 10k reads. <em>Try to calculate the mean and median number of reads on your own.</em></p>
<p>The last line of code above can be used to add a new column containing the total read count to the metadata data.frame. Similarly, sample_data(ps)$total_reads &lt;- sample_sums(ps) would add this information to the phyloseq object itself (as a new sample_data variable).</p>
<p><br></p>
</div>
<div id="examining-the-otu-table" class="section level1">
<h1>Examining the OTU table</h1>
<pre class="r"><code>ntaxa(ps)</code></pre>
<pre><code>## [1] 232</code></pre>
<pre class="r"><code>head(taxa_names(ps))</code></pre>
<pre><code>## [1] &quot;ASV1&quot; &quot;ASV2&quot; &quot;ASV3&quot; &quot;ASV4&quot; &quot;ASV5&quot; &quot;ASV6&quot;</code></pre>
<pre class="r"><code>head(taxa_sums(ps))</code></pre>
<pre><code>##  ASV1  ASV2  ASV3  ASV4  ASV5  ASV6 
## 14148  9898  8862  7935  5880  5469</code></pre>
<pre class="r"><code>(asv_tab &lt;- data.frame(otu_table(ps)[1:5, 1:5]))</code></pre>
<pre><code>##        ASV1 ASV2 ASV3 ASV4 ASV5
## F3D0    579  345  449  430  154
## F3D1    405  353  231   69  140
## F3D141  444  362  345  502  189
## F3D142  289  304  158  164  180
## F3D143  228  176  204  231  130</code></pre>
<ul>
<li>Phyloseq allows you to easily:
<ul>
<li>Obtain a count of the number of taxa</li>
<li>Access their names (e.g. ASV1, ASV2, …)</li>
<li>Get a count of each ASV summed over all samples</li>
<li>Extract the OTU table as a data.frame</li>
</ul></li>
</ul>
<p><br></p>
</div>
<div id="examining-the-taxonomy" class="section level1">
<h1>Examining the taxonomy</h1>
<pre class="r"><code>rank_names(ps)</code></pre>
<pre><code>## [1] &quot;Kingdom&quot; &quot;Phylum&quot;  &quot;Class&quot;   &quot;Order&quot;   &quot;Family&quot;  &quot;Genus&quot;   &quot;Species&quot;</code></pre>
<pre class="r"><code>head(tax_table(ps))</code></pre>
<pre><code>## Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
##      Kingdom    Phylum          Class         Order          
## ASV1 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV2 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV3 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV4 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV5 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV6 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
##      Family           Genus         Species
## ASV1 &quot;Muribaculaceae&quot; NA            NA     
## ASV2 &quot;Muribaculaceae&quot; NA            NA     
## ASV3 &quot;Muribaculaceae&quot; NA            NA     
## ASV4 &quot;Muribaculaceae&quot; NA            NA     
## ASV5 &quot;Bacteroidaceae&quot; &quot;Bacteroides&quot; NA     
## ASV6 &quot;Muribaculaceae&quot; NA            NA</code></pre>
<pre class="r"><code>head(tax_table(ps)[, 2])</code></pre>
<pre><code>## Taxonomy Table:     [6 taxa by 1 taxonomic ranks]:
##      Phylum         
## ASV1 &quot;Bacteroidetes&quot;
## ASV2 &quot;Bacteroidetes&quot;
## ASV3 &quot;Bacteroidetes&quot;
## ASV4 &quot;Bacteroidetes&quot;
## ASV5 &quot;Bacteroidetes&quot;
## ASV6 &quot;Bacteroidetes&quot;</code></pre>
<pre class="r"><code>table(tax_table(ps)[, 2])</code></pre>
<pre><code>## 
##      Actinobacteria       Bacteroidetes       Cyanobacteria 
##                   6                  20                   3 
## Deinococcus-Thermus  Epsilonbacteraeota          Firmicutes 
##                   1                   1                 185 
##     Patescibacteria      Proteobacteria         Tenericutes 
##                   2                   7                   6 
##     Verrucomicrobia 
##                   1</code></pre>
<pre class="r"><code>(tax_tab &lt;- data.frame(tax_table(ps)[50:55, ]))</code></pre>
<pre><code>##        Kingdom             Phylum           Class             Order
## ASV50 Bacteria         Firmicutes      Clostridia     Clostridiales
## ASV51 Bacteria         Firmicutes      Clostridia     Clostridiales
## ASV52 Bacteria         Firmicutes      Clostridia     Clostridiales
## ASV53 Bacteria Epsilonbacteraeota Campylobacteria Campylobacterales
## ASV54 Bacteria         Firmicutes      Clostridia     Clostridiales
## ASV55 Bacteria         Firmicutes      Clostridia     Clostridiales
##                  Family                   Genus Species
## ASV50   Lachnospiraceae           Acetatifactor    &lt;NA&gt;
## ASV51   Ruminococcaceae     Ruminiclostridium_5    &lt;NA&gt;
## ASV52   Lachnospiraceae Lachnospiraceae_UCG-001    &lt;NA&gt;
## ASV53 Helicobacteraceae            Helicobacter  pylori
## ASV54       Family_XIII                    &lt;NA&gt;    &lt;NA&gt;
## ASV55   Ruminococcaceae     Ruminiclostridium_5    &lt;NA&gt;</code></pre>
<p>Here we can see that we have information on 7 taxonomic ranks from Kingdom to Species. We can easily access specific components of this object to learn more about the classifications. For example, we see that the vast majority of ASVs are classified as Firmicutes. This is in line with our expectations. Conducting such assessments may help you identify potential sequencing errors that made it through the denoising pipeline (i.e. those not assigned to a Kingdom) or to understand the proportion of sequences classified at lower levels (i.e. genus or species).</p>
<p>One could also swap out this taxonomy file for another…say using the <a href="https://rdrr.io/bioc/DECIPHER/man/IdTaxa.html">IDTAXA</a> function in the <a href="http://www.bioconductor.org/packages/release/bioc/html/DECIPHER.html">DECIPHER</a> package or an alternative reference database (i.e. Silva or Greengenes). <em>I will let you look into this on your own!</em></p>
<p><br></p>
</div>
<div id="examining-the-reference-sequences" class="section level1">
<h1>Examining the reference sequences</h1>
<p>Storing the reference sequences with your phyloseq object is critical of you rename the ASV names to ASV1, ASV2, …
This will allow you to communicate the information on these ASVs directly (i.e. you can provide the exact sequence variant information). This information is also required to build a phylogenetic tree or BLAST the sequences against the NCBI database for example. <strong>In short, always include these in the phyloseq object.</strong></p>
<p>Below we see that these sequences are stored as a DNAStringSet. The refseq command returns the ASV number, sequence length, and amplicon sequence for each ASV. The function <a href="https://rdrr.io/bioc/dada2/man/nwhamming.html">dada2::nwhamming</a> is calculating the Hamming distance of two sequences after alignment. We will discuss more about this in class. We can also pull out the component and store it as a data.frame.</p>
<pre class="r"><code>head(refseq(ps))</code></pre>
<pre><code>##   A DNAStringSet instance of length 6
##     width seq                                          names               
## [1]   252 TACGGAGGATGCGAGCGTTAT...AAGTGTGGGTATCGAACAGG ASV1
## [2]   252 TACGGAGGATGCGAGCGTTAT...AAGCGTGGGTATCGAACAGG ASV2
## [3]   252 TACGGAGGATGCGAGCGTTAT...AAGCGTGGGTATCGAACAGG ASV3
## [4]   252 TACGGAGGATGCGAGCGTTAT...AAGTGCGGGGATCGAACAGG ASV4
## [5]   253 TACGGAGGATCCGAGCGTTAT...AAGTGTGGGTATCAAACAGG ASV5
## [6]   252 TACGGAGGATGCGAGCGTTAT...AAGTGCGGGGATCAAACAGG ASV6</code></pre>
<pre class="r"><code>dada2::nwhamming(as.vector(refseq(ps)[1]), as.vector(refseq(ps)[2]))</code></pre>
<pre><code>## [1] 20</code></pre>
<pre class="r"><code>(ref_tab &lt;- data.frame(head(refseq(ps))))</code></pre>
<pre><code>##                                                                                                                                                                                                                                                   head.refseq.ps..
## ASV1  TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGAAGATCAAGTCAGCGGTAAAATTGAGAGGCTCAACCTCTTCGAGCCGTTGAAACTGGTTTTCTTGAGTGAGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCTCAACTGACGCTCATGCACGAAAGTGTGGGTATCGAACAGG
## ASV2  TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGACTCTCAAGTCAGCGGTCAAATCGCGGGGCTCAACCCCGTTCCGCCGTTGAAACTGGGAGCCTTGAGTGCGCGAGAAGTAGGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCCTACCGGCGCGCAACTGACGCTCATGCACGAAAGCGTGGGTATCGAACAGG
## ASV3  TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGGCTGTTAAGTCAGCGGTCAAATGTCGGGGCTCAACCCCGGCCTGCCGTTGAAACTGGCGGCCTCGAGTGGGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCCCGACTGACGCTGAGGCACGAAAGCGTGGGTATCGAACAGG
## ASV4  TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGGCTTTTAAGTCAGCGGTAAAAATTCGGGGCTCAACCCCGTCCGGCCGTTGAAACTGGGGGCCTTGAGTGGGCGAGAAGAAGGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCCTTCCGGCGCCCTACTGACGCTGAGGCACGAAAGTGCGGGGATCGAACAGG
## ASV5 TACGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGGTGGATTGTTAAGTCAGTTGTGAAAGTTTGCGGCTCAACCGTAAAATTGCAGTTGAAACTGGCAGTCTTGAGTACAGTAGAGGTGGGCGGAATTCGTGGTGTAGCGGTGAAATGCTTAGATATCACGAAGAACTCCGATTGCGAAGGCAGCTCACTGGACTGCAACTGACACTGATGCTCGAAAGTGTGGGTATCAAACAGG
## ASV6  TACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGCGGCCTGCCAAGTCAGCGGTAAAATTGCGGGGCTCAACCCCGTACAGCCGTTGAAACTGCCGGGCTCGAGTGGGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACCCCGATTGCGAAGGCAGCATACCGGCGCCCTACTGACGCTGAGGCACGAAAGTGCGGGGATCAAACAGG</code></pre>
<p><br></p>
</div>
<div id="accessing-the-phylogenetic-tree" class="section level1">
<h1>Accessing the phylogenetic tree</h1>
<p>We did not generate a phylogenetic tree during the DADA2 tutorial in the interest of time. However, phyloseq has many excellent tools for working with and visualizing trees. I recommend you take a look at these tutorials below for some examples.</p>
<ul>
<li><a href="https://joey711.github.io/phyloseq/preprocess.html" class="uri">https://joey711.github.io/phyloseq/preprocess.html</a></li>
<li><a href="https://joey711.github.io/phyloseq/plot_tree-examples.html" class="uri">https://joey711.github.io/phyloseq/plot_tree-examples.html</a></li>
</ul>
<p>Ben Callahan’s <a href="https://f1000research.com/articles/5-1492">F1000</a> paper demonstrates a complete analysis workflow in R including the construction of a de-novo phylogenetic tree. I highly recommned you take a look at this paper.</p>
<p><br></p>
</div>
<div id="agglomerating-and-subsetting-taxa" class="section level1">
<h1>Agglomerating and subsetting taxa</h1>
<p>Often times we may want to agglomerate taxa to a specific taxonomic rank for analysis. Or we may want to work with a given subset of taxa. We can perform these operations in phyloseq with the tax_glom and subset_taxa functions.</p>
<pre class="r"><code>(ps_phylum &lt;- tax_glom(ps, &quot;Phylum&quot;))</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 10 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 10 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 10 reference sequences ]</code></pre>
<pre class="r"><code>taxa_names(ps_phylum)</code></pre>
<pre><code>##  [1] &quot;ASV1&quot;   &quot;ASV11&quot;  &quot;ASV19&quot;  &quot;ASV53&quot;  &quot;ASV67&quot;  &quot;ASV90&quot;  &quot;ASV107&quot;
##  [8] &quot;ASV109&quot; &quot;ASV189&quot; &quot;ASV191&quot;</code></pre>
<pre class="r"><code>taxa_names(ps_phylum) &lt;- tax_table(ps_phylum)[, 2]
taxa_names(ps_phylum)</code></pre>
<pre><code>##  [1] &quot;Bacteroidetes&quot;       &quot;Firmicutes&quot;          &quot;Tenericutes&quot;        
##  [4] &quot;Epsilonbacteraeota&quot;  &quot;Actinobacteria&quot;      &quot;Patescibacteria&quot;    
##  [7] &quot;Proteobacteria&quot;      &quot;Deinococcus-Thermus&quot; &quot;Cyanobacteria&quot;      
## [10] &quot;Verrucomicrobia&quot;</code></pre>
<pre class="r"><code>otu_table(ps_phylum)[1:5, c(1:3, 5, 7)]</code></pre>
<pre><code>## OTU Table:          [5 taxa and 5 samples]
##                      taxa are columns
##        Bacteroidetes Firmicutes Tenericutes Actinobacteria Proteobacteria
## F3D0            3708       2620         151             27             12
## F3D1            1799       3011         157              3             16
## F3D141          3437       1370          35             16              0
## F3D142          2003        452          33             28              0
## F3D143          1816        655          34             10              0</code></pre>
<p>Here we are agglomerating the counts to the Phylum-level and then renaming the ASVs to make them more descriptive.
We can see that we have 10 Phyla. The ASV information (i.e. refseq and taxonomy for one of the ASVs in each Phylum) gets carried along for the ride (we can typically ignore this or you can remove these components if you prefer).</p>
<p><br></p>
<div id="we-can-also-subset-taxa" class="section level4">
<h4>We can also subset taxa…</h4>
<pre class="r"><code>(ps_bacteroides &lt;- subset_taxa(ps, Genus == &quot;Bacteroides&quot;))</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 3 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 3 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 3 reference sequences ]</code></pre>
<pre class="r"><code>tax_table(ps_bacteroides)</code></pre>
<pre><code>## Taxonomy Table:     [3 taxa by 7 taxonomic ranks]:
##        Kingdom    Phylum          Class         Order          
## ASV5   &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV80  &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
## ASV163 &quot;Bacteria&quot; &quot;Bacteroidetes&quot; &quot;Bacteroidia&quot; &quot;Bacteroidales&quot;
##        Family           Genus         Species   
## ASV5   &quot;Bacteroidaceae&quot; &quot;Bacteroides&quot; NA        
## ASV80  &quot;Bacteroidaceae&quot; &quot;Bacteroides&quot; &quot;vulgatus&quot;
## ASV163 &quot;Bacteroidaceae&quot; &quot;Bacteroides&quot; &quot;vulgatus&quot;</code></pre>
<pre class="r"><code>prune_taxa(taxa_sums(ps) &gt; 100, ps) </code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 99 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 99 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 99 reference sequences ]</code></pre>
<pre class="r"><code>filter_taxa(ps, function(x) sum(x &gt; 10) &gt; (0.1*length(x)), TRUE)   </code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 135 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 135 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 135 reference sequences ]</code></pre>
<ul>
<li>With the above commands we can quickly see that we have:
<ul>
<li>A total of 3 ASVs classified as Bacteroides</li>
<li>A total of 99 ASVs seen at least 100 times across all samples</li>
<li>A total of 135 taxa seen at least 10 times in at least 10% of samples</li>
</ul></li>
</ul>
<p><strong>This highlights how we might use phyloseq as a tool to filter taxa prior to statistical analysis.</strong></p>
<p><br></p>
</div>
</div>
<div id="subsetting-samples-and-tranforming-counts" class="section level1">
<h1>Subsetting samples and tranforming counts</h1>
<p>Phyloseq can also be used to subset all the individual components based on sample metadata information. <strong>This would take a fair bit of work to do properly if we were working with each individual component…and not with phyloseq.</strong> Below we subset the early stool samples. Then we generate an object that includes only samples with &gt; 5,000 total reads.</p>
<pre class="r"><code>ps_early &lt;- subset_samples(ps, When == &quot;Early&quot;)
(ps_early = prune_taxa(taxa_sums(ps_early) &gt; 0, ps_early))</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 183 taxa and 9 samples ]
## sample_data() Sample Data:       [ 9 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 183 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 183 reference sequences ]</code></pre>
<pre class="r"><code>sample_data(ps_early)$When</code></pre>
<pre><code>## [1] &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot; &quot;Early&quot;</code></pre>
<pre class="r"><code>sort(sample_sums(ps))</code></pre>
<pre><code>## F3D143 F3D142 F3D144   F3D5 F3D146   F3D7 F3D150   F3D8 F3D141   F3D1 
##   2518   2521   3488   3716   3879   4217   4240   4547   4863   5017 
##   F3D3 F3D145   F3D9   F3D0   F3D6 F3D148 F3D149 F3D147   F3D2 
##   5491   5820   6015   6528   6679   9935  10653  13006  16835</code></pre>
<pre class="r"><code>(ps_reads_GT_5k = prune_samples(sample_sums(ps) &gt; 5000, ps))</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 232 taxa and 10 samples ]
## sample_data() Sample Data:       [ 10 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 232 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 232 reference sequences ]</code></pre>
<pre class="r"><code>sort(sample_sums(ps_reads_GT_5k))</code></pre>
<pre><code>##   F3D1   F3D3 F3D145   F3D9   F3D0   F3D6 F3D148 F3D149 F3D147   F3D2 
##   5017   5491   5820   6015   6528   6679   9935  10653  13006  16835</code></pre>
<p><br></p>
<p>Counts can be converted to relative abundances (e.g. total sum scaling) using the transform_sample_counts function. They can also be subsampled/rarified using the rarefy_even_depth function. However, subsampling to account for differences in sequencing depth acorss samples has important limitations. See the papers below for a more in-depth discussion.</p>
<ul>
<li><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531">McMurdie and Holmes, Waste Not, Want Not: Why Rarefying Microbiome Data Is Inadmissible</a></li>
<li><a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-017-0237-y">Weiss et. al., Normalization and microbial differential abundance strategies depend upon data characteristics</a></li>
</ul>
<p><br></p>
<pre class="r"><code>ps_relabund &lt;- transform_sample_counts(ps, function(x) x / sum(x))
otu_table(ps_relabund)[1:5, 1:5]</code></pre>
<pre><code>## OTU Table:          [5 taxa and 5 samples]
##                      taxa are columns
##              ASV1       ASV2       ASV3       ASV4       ASV5
## F3D0   0.08869485 0.05284926 0.06878064 0.06587010 0.02359069
## F3D1   0.08072553 0.07036077 0.04604345 0.01375324 0.02790512
## F3D141 0.09130167 0.07443965 0.07094386 0.10322846 0.03886490
## F3D142 0.11463705 0.12058707 0.06267354 0.06505355 0.07140024
## F3D143 0.09054805 0.06989674 0.08101668 0.09173948 0.05162828</code></pre>
<pre class="r"><code>(ps_rare &lt;- rarefy_even_depth(ps, sample.size = 4000, rngseed = 123, replace = FALSE))</code></pre>
<pre><code>## `set.seed(123)` was used to initialize repeatable random subsampling.</code></pre>
<pre><code>## Please record this for your records so others can reproduce.</code></pre>
<pre><code>## Try `set.seed(123); .Random.seed` for the full vector</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## 5 samples removedbecause they contained fewer reads than `sample.size`.</code></pre>
<pre><code>## Up to first five removed samples are:</code></pre>
<pre><code>## F3D142F3D143F3D144F3D146F3D5</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## 15OTUs were removed because they are no longer 
## present in any sample after random subsampling</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 217 taxa and 14 samples ]
## sample_data() Sample Data:       [ 14 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 217 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 217 reference sequences ]</code></pre>
<pre class="r"><code>sample_sums(ps_rare)</code></pre>
<pre><code>##   F3D0   F3D1 F3D141 F3D145 F3D147 F3D148 F3D149 F3D150   F3D2   F3D3 
##   4000   4000   4000   4000   4000   4000   4000   4000   4000   4000 
##   F3D6   F3D7   F3D8   F3D9 
##   4000   4000   4000   4000</code></pre>
<p><br></p>
</div>
<div id="example-analytic-and-graphical-capabilities" class="section level1">
<h1>Example analytic and graphical capabilities</h1>
<p>Phyloseq has an extensive list of functions for processing and analyzing microbiome data. I recommend you view the tutorial section on the phyloseq home page to get a feel for all that phyloseq can do. Below are just a few quick examples. We will get more into these types of analyses in subsequent sessions.</p>
<div id="alpha-diversity" class="section level3">
<h3>Alpha-diversity</h3>
<p>Below we will receive a warning that our data does not contain any singletons and that the results of richness estimates are probably unreliable. This is an important point and we will delve into this issue more in the next session. For now, you can go ahead and ignore the warning.</p>
<pre class="r"><code>head(estimate_richness(ps))</code></pre>
<pre><code>## Warning in estimate_richness(ps): The data you have provided does not have
## any singletons. This is highly suspicious. Results of richness
## estimates (for example) are probably unreliable, or wrong, if you have already
## trimmed low-abundance taxa from the data.
## 
## We recommended that you find the un-trimmed data and retry.</code></pre>
<pre><code>##        Observed Chao1 se.chao1 ACE   se.ACE  Shannon   Simpson InvSimpson
## F3D0        106   106        0 106 4.539138 3.865881 0.9644889   28.16024
## F3D1        100   100        0 100 4.208325 3.993196 0.9709838   34.46347
## F3D141       74    74        0  74 3.878214 3.428895 0.9501123   20.04502
## F3D142       48    48        0  48 3.388092 3.117940 0.9386949   16.31185
## F3D143       56    56        0  56 3.543102 3.292717 0.9464422   18.67141
## F3D144       47    47        0  47 3.135249 2.994201 0.9309895   14.49054
##           Fisher
## F3D0   17.973004
## F3D1   17.696857
## F3D141 12.383762
## F3D142  8.412094
## F3D143 10.148818
## F3D144  7.678694</code></pre>
<pre class="r"><code>(p &lt;- plot_richness(ps, x = &quot;When&quot;, color = &quot;When&quot;, measures = c(&quot;Observed&quot;, &quot;Shannon&quot;)))</code></pre>
<pre><code>## Warning in estimate_richness(physeq, split = TRUE, measures = measures): The data you have provided does not have
## any singletons. This is highly suspicious. Results of richness
## estimates (for example) are probably unreliable, or wrong, if you have already
## trimmed low-abundance taxa from the data.
## 
## We recommended that you find the un-trimmed data and retry.</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/adiv-1.png" width="672" /></p>
<pre class="r"><code>p + labs(x = &quot;&quot;, y = &quot;Alpha Diversity Measure\n&quot;) + 
  theme_bw()</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/adiv-2.png" width="672" /></p>
</div>
<div id="beta-diversity-ordination" class="section level3">
<h3>Beta-diversity ordination</h3>
<pre class="r"><code>ps_rare_bray &lt;- ordinate(ps_rare, &quot;NMDS&quot;, &quot;bray&quot;)</code></pre>
<pre><code>## Square root transformation
## Wisconsin double standardization
## Run 0 stress 0.08484704 
## Run 1 stress 0.08484704 
## ... New best solution
## ... Procrustes: rmse 2.497137e-06  max resid 5.691675e-06 
## ... Similar to previous best
## Run 2 stress 0.09657264 
## Run 3 stress 0.08484704 
## ... Procrustes: rmse 7.186183e-07  max resid 1.423558e-06 
## ... Similar to previous best
## Run 4 stress 0.08484704 
## ... Procrustes: rmse 3.303025e-06  max resid 7.565974e-06 
## ... Similar to previous best
## Run 5 stress 0.1744901 
## Run 6 stress 0.08484704 
## ... Procrustes: rmse 1.008148e-06  max resid 2.038791e-06 
## ... Similar to previous best
## Run 7 stress 0.08484704 
## ... Procrustes: rmse 1.776536e-06  max resid 3.520974e-06 
## ... Similar to previous best
## Run 8 stress 0.09657264 
## Run 9 stress 0.08484704 
## ... Procrustes: rmse 8.550518e-07  max resid 1.794331e-06 
## ... Similar to previous best
## Run 10 stress 0.08484704 
## ... Procrustes: rmse 1.376679e-06  max resid 2.816876e-06 
## ... Similar to previous best
## Run 11 stress 0.08484704 
## ... Procrustes: rmse 4.702272e-06  max resid 8.17489e-06 
## ... Similar to previous best
## Run 12 stress 0.08484704 
## ... New best solution
## ... Procrustes: rmse 2.157179e-07  max resid 4.2813e-07 
## ... Similar to previous best
## Run 13 stress 0.08484704 
## ... Procrustes: rmse 1.726469e-06  max resid 3.270828e-06 
## ... Similar to previous best
## Run 14 stress 0.08484704 
## ... Procrustes: rmse 1.055175e-06  max resid 2.649077e-06 
## ... Similar to previous best
## Run 15 stress 0.09657265 
## Run 16 stress 0.1751066 
## Run 17 stress 0.08484704 
## ... Procrustes: rmse 6.953774e-07  max resid 1.374792e-06 
## ... Similar to previous best
## Run 18 stress 0.09584961 
## Run 19 stress 0.08484704 
## ... Procrustes: rmse 5.428812e-06  max resid 1.248684e-05 
## ... Similar to previous best
## Run 20 stress 0.1795526 
## *** Solution reached</code></pre>
<pre class="r"><code>plot_ordination(ps_rare, ps_rare_bray, type=&quot;samples&quot;, color=&quot;When&quot;) + geom_point(size = 3) </code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/bdiv-1.png" width="672" /></p>
</div>
<div id="stacked-bar-plots" class="section level3">
<h3>Stacked bar plots</h3>
<pre class="r"><code>plot_bar(ps, fill=&quot;Phylum&quot;)</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/taxplot-1.png" width="672" /></p>
<pre class="r"><code>plot_bar(ps_relabund, fill=&quot;Phylum&quot;) + 
  geom_bar(aes(color = Phylum, fill = Phylum), stat=&quot;identity&quot;, position=&quot;stack&quot;) +
  labs(x = &quot;&quot;, y = &quot;Relative Abundance\n&quot;) +
  theme(panel.background = element_blank())</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/taxplot-2.png" width="672" /></p>
</div>
<div id="heatmaps" class="section level3">
<h3>Heatmaps</h3>
<pre class="r"><code>(ps_fam &lt;- tax_glom(ps, &quot;Family&quot;))</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 33 taxa and 19 samples ]
## sample_data() Sample Data:       [ 19 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 33 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 33 reference sequences ]</code></pre>
<pre class="r"><code>(ps_fam_rare &lt;- rarefy_even_depth(ps_fam, sample.size = 4000, rngseed = 123, replace = FALSE))</code></pre>
<pre><code>## `set.seed(123)` was used to initialize repeatable random subsampling.</code></pre>
<pre><code>## Please record this for your records so others can reproduce.</code></pre>
<pre><code>## Try `set.seed(123); .Random.seed` for the full vector</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## 5 samples removedbecause they contained fewer reads than `sample.size`.</code></pre>
<pre><code>## Up to first five removed samples are:</code></pre>
<pre><code>## F3D142F3D143F3D144F3D146F3D5</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## 9OTUs were removed because they are no longer 
## present in any sample after random subsampling</code></pre>
<pre><code>## ...</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 24 taxa and 14 samples ]
## sample_data() Sample Data:       [ 14 samples by 4 sample variables ]
## tax_table()   Taxonomy Table:    [ 24 taxa by 7 taxonomic ranks ]
## refseq()      DNAStringSet:      [ 24 reference sequences ]</code></pre>
<pre class="r"><code>plot_heatmap(ps_fam_rare, sample.label = &quot;When&quot;, taxa.label = &quot;Family&quot;)</code></pre>
<pre><code>## Warning: Transformation introduced infinite values in discrete y-axis</code></pre>
<p><img src="/post/2019-07-28-introduction-to-phyloseq_files/figure-html/heatmap-1.png" width="672" /></p>
</div>
</div>

    </div>

    

<div class="article-tags">
  
  <a class="badge badge-light" href="/tags/microbiome/">Microbiome</a>
  
  <a class="badge badge-light" href="/tags/r/">R</a>
  
  <a class="badge badge-light" href="/tags/phyloseq/">Phyloseq</a>
  
  <a class="badge badge-light" href="/tags/dada2/">DADA2</a>
  
</div>



    
      








  





  
  
  
    
  
  
  <div class="media author-card" itemscope itemtype="http://schema.org/Person">
    
      
      <img class="portrait mr-3" src="/authors/admin/avatar_hu87de4e6533f526eb4329470784cb7712_907390_250x250_fill_q90_lanczos_center.jpg" itemprop="image" alt="Avatar">
    

    <div class="media-body">
      <h5 class="card-title" itemprop="name"><a href="/">Nicholas Ollberding</a></h5>
      <h6 class="card-subtitle">Associate Professor of Pediatrics and Nutrition</h6>
      
      <ul class="network-icon" aria-hidden="true">
        
          
          
          
            
          
          
          
          
          
          <li>
            <a itemprop="sameAs" href="mailto:nicholas.ollberding@cchmc.org" >
              <i class="fas fa-envelope"></i>
            </a>
          </li>
        
          
          
          
            
          
          
          
          
          
            
          
          <li>
            <a itemprop="sameAs" href="https://twitter.com/Nick10243" target="_blank" rel="noopener">
              <i class="fab fa-twitter"></i>
            </a>
          </li>
        
          
          
          
          
          
          
          
            
          
          <li>
            <a itemprop="sameAs" href="https://scholar.google.com/citations?hl=en&amp;user=k9mLyI0AAAAJ" target="_blank" rel="noopener">
              <i class="ai ai-google-scholar"></i>
            </a>
          </li>
        
          
          
          
            
          
          
          
          
          
            
          
          <li>
            <a itemprop="sameAs" href="https://github.com/Nick243" target="_blank" rel="noopener">
              <i class="fab fa-github"></i>
            </a>
          </li>
        
          
          
          
          
          
          
          
            
          
          <li>
            <a itemprop="sameAs" href="https://www.researchgate.net/profile/Nicholas_Ollberding" target="_blank" rel="noopener">
              <i class="ai ai-researchgate"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>



      
      
      <div class="article-widget">
        <div class="hr-light"></div>
        <h3>Related</h3>
        <ul>
          
          <li><a href="/talk/test-talk/">Introduction to Denoising Amplicon Sequence Data in R</a></li>
          
          <li><a href="/talk/into-mb-stats/">Introduction to Microbiome Data Analysis </a></li>
          
          <li><a href="/post/introduction-to-the-statistical-analysis-of-microbiome-data-in-r/">Introduction to the Statistical Analysis of Microbiome Data in R</a></li>
          
          <li><a href="/talk/dhc/">Recent Advancements in the Statistical Analysis of Microbial Metagenomic Sequence Data</a></li>
          
          <li><a href="/publication/candle-study/">Urinary Microbiota Associated with Preterm Birth: Results from the Conditions Affecting Neurocognitive Development and Learning in Early Childhood (CANDLE) Study</a></li>
          
        </ul>
      </div>
      
    

    

    


  </div>
</article>

      

    
    

    
    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.4/imagesloaded.pkgd.min.js" integrity="sha256-lqvxZrPLtfffUl2G/e7szqSvPBILGbwmsGE1MKlOi0Q=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.6/isotope.pkgd.min.js" integrity="sha256-CBrpuqrMhXwcLLUd5tvQ4euBHCdh7wGlDfNz8vbu/iI=" crossorigin="anonymous"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>

      

      
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/highlight.min.js" integrity="sha256-aYTdUrn6Ow1DDgh5JTc3aDGnnju48y/1c8s1dgkYPQ8=" crossorigin="anonymous"></script>
        
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.6/languages/r.min.js"></script>
        
      

      
      
    

    
    
      <script src="https://cdnjs.cloudflare.com/ajax/libs/leaflet/1.2.0/leaflet.js" integrity="sha512-lInM/apFSqyy1o6s89K4iQUKg6ppXEgsVxT35HbzUupEVRh2Eu9Wdl4tHj7dZO0s1uvplcYGmt3498TtHq+log==" crossorigin="anonymous"></script>
    

    
    
    

    
    
    <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script>
      const search_index_filename = "/index.json";
      const i18n = {
        'placeholder': "Search...",
        'results': "results found",
        'no_results': "No results found"
      };
      const content_type = {
        'post': "Posts",
        'project': "Projects",
        'publication' : "Publications",
        'talk' : "Talks"
        };
    </script>
    

    
    

    
    
    <script id="search-hit-fuse-template" type="text/x-template">
      <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
      </div>
    </script>
    

    
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.2.1/fuse.min.js" integrity="sha256-VzgmKYmhsGNNN4Ph1kMW+BjoYJM2jV5i4IlFoeZA9XI=" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>
    

    
    

    
    
    
    
    
    
    
    
    
      
    
    
    
    
    <script src="/js/academic.min.ac3764a0b2acc1c82ace08e756adafc1.js"></script>

    






  
  <div class="container">
    <footer class="site-footer">
  

  <p class="powered-by">
    

    Powered by the
    <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
    <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

    
    <span class="float-right" aria-hidden="true">
      <a href="#" id="back_to_top">
        <span class="button_icon">
          <i class="fas fa-chevron-up fa-2x"></i>
        </span>
      </a>
    </span>
    
  </p>
</footer>

  </div>
  

  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        <pre><code class="tex hljs"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

</body>
</html>
